### ISSUE TITLE
router guard and async

### ISSUE THREAD

**Issue by Manborough** [ItalyPaleAle/svelte-spa-router#125]
Is there a particular way I need to use async functions within the wrap function?
I'm trying to check authorisation from my API roughly like this.

```  const routes = {
    '/login': Login,
    '/facility': wrap(
      Facility,
      {},
      async (detail) => {
        const result = await authQuery()
        return await result
      }
      ),
    '*': Login,
  }
```

but I'm finding the route loads before the results from the api return.

**Comment by Doudou8** [comment]
I also have this issue. I need to verify that the user is connected with the firebase API.

**Comment by BogdanDarius** [comment]
Hi!
You could specify in routes config a wrapper component:
```js
let routes;
routes = new Map();
routes.set("/protected-route", MyWrapper);
```

And in `MyWrapper` you could do something like this:
```js
{#await isAuthenticated()}
// Show progress
{:then value}
// Load component with Loadable or svelte:component
  <Loadable loader={AccessGranted} />
{:catch error}
 // Show error
{/await}

**Comment by Dan1ve** [comment]
I have the same issue, and I feel like the suggestion by @BogdanDarius isn't general-purpose enough to (easily) work on many pages (please correct me if I misunderstood something).
Simply passing an `async` function to `wrap` would be really nice!

**Comment by BogdanDarius** [comment]
> I have the same issue, and I feel like the suggestion by @BogdanDarius isn't general-purpose enough to (easily) work on many pages (please correct me if I misunderstood something).
> Simply passing an `async` function to `wrap` would be really nice!
Promise and lazy routing is on the roadmap.
I encountered same issue while I wanted to implement code splitting.

**Comment by BogdanDarius** [comment]
> I have the same issue, and I feel like the suggestion by @BogdanDarius isn't general-purpose enough to (easily) work on many pages (please correct me if I misunderstood something).
> Simply passing an `async` function to `wrap` would be really nice!

Promise and lazy routing is on the roadmap.
I encountered same issue while I wanted to implement code splitting.

**Comment by BogdanDarius** [comment]
> I have the same issue, and I feel like the suggestion by @BogdanDarius isn't general-purpose enough to (easily) work on many pages (please correct me if I misunderstood something).
> Simply passing an `async` function to `wrap` would be really nice!

Promise and lazy routing are on the roadmap.
I encountered same issue while I wanted to implement code splitting.

**Comment by Dan1ve** [comment]
That's great to her @BogdanDarius , really looking forward to this! Let me know if I can help in any way.

**Comment by hollandzone** [comment]
Encountering this issue with firebase authentication as well, would love to see this feature implemented.

**Comment by Dan1ve** [comment]
FWIW, I now call a `redirectIfNotAuthenticated()` method at the top of every `<script>` tag of protected svelte files. This method checks the authentication against the server and, if necessary, redirects to the login page (setting the current page as url argument for redirect back after login). Although not an ideal solution, this seems to work fine as a workaround.

**Comment by ItalyPaleAle** [comment]
I do the same as @Dan1ve in my own code.

If every single route requires authentication, I check it in a function that's called at the beginning of the page's execution, in the main.js file (outside of the Svelte app)

**Comment by ItalyPaleAle** [comment]
@Manborough support for async functions in route pre-conditions has been added to 3.0 (WIP branch) and will be released with version 3.0 soon 🚀
